File: src/main/java/dev/mazurkiewicz/quizer/solution/ValidAnswerType.java
package dev.mazurkiewicz.quizer.solution;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target({ElementType.METHOD, ElementType.FIELD})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = AnswerTypeValidator.class)
public @interface ValidAnswerType {

    String message() default "{dev.mazurkiewicz.quizer.request.validAnswerType.message}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: src/main/kotlin/pl/quizwiz/question/application/ValidAnswerType.kt
package pl.quizwiz.question.application

import jakarta.validation.Constraint
import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import jakarta.validation.Payload
import pl.quizwiz.question.domain.model.AnswerType
import kotlin.reflect.KClass


@Constraint(validatedBy = [AnswerTypeValidator::class])
@Retention(AnnotationRetention.RUNTIME)
@Target(
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER,
    AnnotationTarget.FIELD
)
annotation class ValidAnswerType(
    val message: String = "{dev.mazurkiewicz.quizer.request.validAnswerType.message}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)

class AnswerTypeValidator : ConstraintValidator<ValidAnswerType, String> {
    override fun isValid(value: String, context: ConstraintValidatorContext): Boolean {
        return try {
            AnswerType.of(value)
            true
        } catch (e: IllegalArgumentException) {
            false
        }
    }
}

